File: tool/bindings_generator/generate_bindings.mjs
// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import * as fs from 'fs';
import * as child_process from 'child_process';
import { createRequire } from 'module';
import * as idl from '@webref/idl';
import * as assumedTypesModule from './assumedTypes.mjs';
import * as translatorModule from './translator.mjs';
import * as util from './util.mjs';

const require = createRequire(import.meta.url);
const Translator = translatorModule.Translator;
const assumedTypesBlob = assumedTypesModule.assumedTypesBlob;

function generateHeader() {
  return (
`// Copyright (c) ${util.year}, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// TODO(joshualitt): Import js types properly.
import 'dart:typed_data';
import 'package:js/js.dart';
`);
}

async function translateIDLs() {
  const files = await idl.listAll();
  const translator = new Translator();
  for (const [shortname, file] of Object.entries(files)) {
    translator.collect(await file.parse());
  }
  return translator.translate();
}

function runDartFormat(filePath) {
  child_process.exec(`dart format ${filePath}`, (error, stdout, stderr) => {
    if (error) {
      console.error(`exec error: ${error}`);
      return;
    }
    console.log(`stdout: ${stdout}`);
    console.error(`stderr: ${stderr}`);
  });
}

async function generateBindings(dir) {
  const filePath = `${dir}/dom.dart`;
  const translation = await translateIDLs();
  console.log(translation);
  fs.writeFileSync(filePath,
`${generateHeader()}
${translation}
${assumedTypesBlob}`);
  runDartFormat(filePath);
}

async function main() {
  const dir = process.argv[2];
  util.ensureDirectoryExists(dir);
  try {
    await generateBindings(dir);
  } catch (e) {
    console.log(e);
    throw e;
  }
}

if (require.main === require.module) {
  main();
}

File: tool/bindings_generator/generate_bindings.dart
// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:js_interop';

import 'translator.dart';
import 'util.dart';
import 'webref_css_api.dart';
import 'webref_idl_api.dart';

/// Generate CSS property names for setting / getting CSS properties in JS.
Future<List<String>> _generateCSSStyleDeclarations() async {
  final cssStyleDeclarations = <String>{};
  final array = objectEntries(await css.listAll().toDart as JSObject);
  for (var i = 0; i < array.length.toDart; i++) {
    final entry = array[i.toJS] as JSArray;
    final data = entry[1.toJS] as CSSEntries;
    final properties = data.properties;
    if (properties != null) {
      for (var j = 0; j < properties.length.toDart; j++) {
        final property = properties[j.toJS] as CSSEntry;
        // There are thre cases for [styleDeclaration]:
        //   1) Length == 1, a single word CSS property.
        //   2) Length == 2, a kebab case property + a camel case property.
        //   3) Length == 3, webkit CSS properties.
        final styleDeclaration = property.styleDeclaration;
        if (styleDeclaration != null) {
          final length = styleDeclaration.length.toDart.toInt();
          if (length < 0 || length > 3) {
            throw Exception('Unexpected style declaration $styleDeclaration');
          }
          // For now we ignore browser specific properties.
          if (length == 3) continue;
          final style =
              (styleDeclaration[(length - 1).toJS] as JSString).toDart;
          if (style.contains('-')) {
            throw Exception('Unexpected style declaration $styleDeclaration');
          }
          cssStyleDeclarations.add(style);
        }
      }
    }
  }
  return cssStyleDeclarations.toList()..sort();
}

Future<TranslationResult> generateBindings(
    String packageRoot, String librarySubDir) async {
  final cssStyleDeclarations = await _generateCSSStyleDeclarations();
  final translator = Translator(librarySubDir, cssStyleDeclarations);
  final array = objectEntries(await idl.parseAll().toDart as JSObject);
  for (var i = 0; i < array.length.toDart; i++) {
    final entry = array[i.toJS] as JSArray;
    final shortname = (entry[0.toJS] as JSString).toDart.kebabToSnake;
    final ast = entry[1.toJS] as JSArray;
    translator.collect(shortname, ast);
  }
  return translator.translate();
}
